pkg/signature/algorithm_registry.go

```diff
@@ -0,0 +1,259 @@
+	hashType crypto.Hash
+
+	// extraKeyParams contains any extra parameters required to check a given public key against this entry.
+	//
+	// The underlying type of these parameters is dependent on the keyType.
+	// For example, ECDSA algorithms will store an elliptic curve here whereas, RSA keys will store the key size.
+	// Algorithms that don't require any extra parameters leave this set to nil.
+	extraKeyParams interface{}
+
+	// flagValue is a string representation of the signature algorithm that follows the naming conventions of CLI
+	// arguments that are used for Sigstore services.
+	flagValue string
+}
+
+func (a algorithmDetailsImpl) GetSignatureAlgorithm() v1.PublicKeyDetails {
+	return a.knownAlgorithm
+}
+
+func (a algorithmDetailsImpl) GetKeyType() PublicKeyType {
+	return a.keyType
+}
+
+func (a algorithmDetailsImpl) GetHashType() crypto.Hash {
+	return a.hashType
+}
+
+func (a algorithmDetailsImpl) GetRSAKeySize() (RSAKeySize, error) {
+	if a.keyType != RSA {
+		return 0, fmt.Errorf("unable to retrieve RSA key size for key type: %T", a.keyType)
+	}
+	rsaKeySize, ok := a.extraKeyParams.(RSAKeySize)
+	if !ok {
+		// This should be unreachable.
+		return 0, fmt.Errorf("unable to retrieve key size for RSA, malformed algorithm details?: %T", a.keyType)
+	}
+	return rsaKeySize, nil
+}
+
+func (a algorithmDetailsImpl) GetECDSACurve() (*elliptic.Curve, error) {
+	if a.keyType != ECDSA {
+		return nil, fmt.Errorf("unable to retrieve ECDSA curve for key type: %T", a.keyType)
+	}
+	ecdsaCurve, ok := a.extraKeyParams.(elliptic.Curve)
+	if !ok {
+		// This should be unreachable.
+		return nil, fmt.Errorf("unable to retrieve curve for ECDSA, malformed algorithm details?: %T", a.keyType)
+	}
+	return &ecdsaCurve, nil
+}
+
+func (a algorithmDetailsImpl) checkKey(pubKey crypto.PublicKey) (bool, error) {
+	switch a.keyType {
+	case RSA:
+		rsaKey, ok := pubKey.(*rsa.PublicKey)
+		if !ok {
+			return false, nil
+		}
+		keySize, err := a.GetRSAKeySize()
+		if err != nil {
+			return false, err
+		}
+		return rsaKey.Size()*8 == int(keySize), nil
+	case ECDSA:
+		ecdsaKey, ok := pubKey.(*ecdsa.PublicKey)
+		if !ok {
+			return false, nil
+		}
+		curve, err := a.GetECDSACurve()
+		if err != nil {
+			return false, err
+		}
+		return ecdsaKey.Curve == *curve, nil
+	case ED25519:
+		_, ok := pubKey.(ed25519.PublicKey)
+		return ok, nil
+	}
+	return false, fmt.Errorf("unrecognized key type: %T", a.keyType)
+}
+
+func (a algorithmDetailsImpl) checkHash(hashType crypto.Hash) bool {
+	return a.hashType == hashType
+}
+
+var algorithmDetails = []algorithmDetailsImpl{
+	{v1.PublicKeyDetails_PKIX_RSA_PKCS1V15_2048_SHA256, RSA, crypto.SHA256, RSAKeySize(2048), "rsa-sign-pkcs1-2048-sha256"},
+	{v1.PublicKeyDetails_PKIX_RSA_PKCS1V15_3072_SHA256, RSA, crypto.SHA256, RSAKeySize(3072), "rsa-sign-pkcs1-3072-sha256"},
+	{v1.PublicKeyDetails_PKIX_RSA_PKCS1V15_4096_SHA256, RSA, crypto.SHA256, RSAKeySize(4096), "rsa-sign-pkcs1-4096-sha256"},
```

Comment:
I think we should add the RSA/PSS options here too.

Reply:
See https://github.com/sigstore/protobuf-specs/blob/main/protos/sigstore_common.proto#L74

Reply:
We should also document we aren't handling the deprecated options, namely a PKCS1v1.5 encoded RSA key, preferring PKIX encoding.

Reply:
Done!